### PR TITLE
Delete extra models from radio when saving models and settings in companion.

### LINKED ISSUE #2817: Model not deleted from radio after saving changes in Companion 2.8
### Is there an existing issue for this problem?

- [X] I have searched the existing issues

### What part of EdgeTX is the focus of this bug?

Companion

### Current Behavior

Deleting a model in Companion 2.8 does not delete the model from the radio when the models and settings are saved to the radio.
If the model is the last one in the list it remains on the radio.
If the model is not the last one in the list then the last model in the list gets duplicated on the radio.

### Expected Behavior

Deleting models in Companion should delete models on the radio when models and settings are written.

### Steps To Reproduce

1. Open Companion 2.8 and connect the radio
2. Read models and settings from the radio
3. Delete a model in Companion
4. Write models and settings to the radio
5. Read models and settings back from the radio
6. The model will re-appear (if it was the last one), or a model will be duplicated if the deleted model was not the last one.

### Version

2.8.0

### Transmitter

Radiomaster TX16S / TX16SMK2

### Anything else?

Tested on Windows 10 and MacOS 12.6.1

### PR DESCRIPTION
**Pull request by philmoz**
Add logic to support deletion of model files from the radio if there are extra models on the radio after saving models and settings in companion.

- only applies to YAML model files that are stored on an SD card in the radio
- this can happen if a model is deleted in companion then saved to the radio
- prompts the user asking if they want to delete all extra models on the radio

Fixes #2817

Summary of changes:
- add logic to check if the radio has more models than the set loaded in companion
- add logic to delete extra models from the radio
- add prompt to ask user if they want to delete extra models

Note: I have not worked out how to add the new prompt strings to the translation files so any advice on this would be appreciated.

### CODE REVIEW

**Comment by elecpower** [comment]
> Note: I have not worked out how to add the new prompt strings to the translation files so any advice on this would be appreciated.

Qt Linguist utility adds to translation files so no additional work for you. @pfeerick regenerates periodically

File: companion/src/storage/sdcard.cpp [hunk 1]
```diff
@@ -82,6 +82,28 @@ bool SdcardFormat::deleteFile(const QString & filename)
   return true;
 }
 
+// Check if the connected radio has more models than the currently loaded set.
+bool SdcardFormat::radioHasExtraModels(const RadioData & radioData)
+{
+  RadioData tmpRadioData;
+  load(tmpRadioData);
+  return tmpRadioData.models.size() > radioData.models.size();
+}
```
Comment (elecpower) [inline companion/src/storage/sdcard.cpp:91]: Noticeable performance hit when lots of models, as all the model files on the radio have to be read, parsed, decoded and loaded into memory variable (including those just written) just to get a count. Why not just count the model files?

File: companion/src/storage/sdcard.cpp [hunk 2]
```diff
@@ -82,6 +82,28 @@ bool SdcardFormat::deleteFile(const QString & filename)
...
+void SdcardFormat::deleteExtraRadioModels(const RadioData & radioData)
+{
+  RadioData tmpRadioData;
+  load(tmpRadioData);
+  if (tmpRadioData.models.size() > radioData.models.size()) {
+    for (int i = radioData.models.size() + 1; i <= tmpRadioData.models.size(); i += 1) {
+      char filename[32];
+      sprintf(filename, "MODELS/model%d.yml", i);
+      deleteFile(filename);
+    }
```
Comment (elecpower) [inline companion/src/storage/sdcard.cpp:103]: Assumes the model files are sequentially numbered from one to the model files count. If model files with names beyond model files count exist they will not be deleted.
If there are any gaps in the sequence of file names a delete error will occur.

**Comment by philmoz** [comment]
Not as simple as I thought. Back to the drawing board.

**Comment by gagarinlg** [comment]
As a general advice, can you please use rebase instead of merge? This keeps the commit logs much cleaner.

Regarding the task, the companion should know the used model files somewhere. Couldn't you just go through the SD cards radio folder and delete files that are not referenced?

**Pull request closed without merging**